### PR TITLE
[ASM] Collect headers on user SDK events

### PR DESCRIPTION
**Pull request by Anilm3**
## Motivation

As part of the process to improve ATO detection, we've decided to start collecting headers on user SDK events:
- When the user generates login success / failure events directly using the SDK.
- When user events are generated through automatic instrumentation and the current mode is `extended`.

## Changes

This PR adds the following new tests:

- tests/appsec/test_automated_login_events.py
  - `Test_Login_Events_Extended::test_login_sdk_success_headers`
  - `Test_Login_Events_Extended::test_login_sdk_failure_headers`

- tests/appsec/test_event_tracking.py
  - `Test_UserLoginSuccessEvent::test_user_login_success_header_collection`
  - `Test_UserLoginFailureEvent::test_user_login_failure_header_collection`

## Workflow

1. ⚠️ Create your PR as draft ⚠️
2. Work on you PR until the CI passes (if something not related to your task is failing, you can ignore it)
3. Mark it as ready for review
    * Test logic is modified? -> Get a review from RFC owner. We're working on refining the `codeowners` file quickly.
    * Framework is modified, or non obvious usage of it -> get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

:rocket: Once your PR is reviewed, you can merge it!

🛟 [#apm-shared-testing](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X) 🛟

## Reviewer checklist

* [ ] [Relevant label](https://github.com/DataDog/system-tests/blob/main/docs/CI/labels.md) (`run-parametric-scenario`, `run-profiling-scenario`...) are presents
* [ ] If PR title starts with `[<language>]`, double-check that only `<language>` is impacted by the change
* [ ] No system-tests internal is modified. Otherwise, I have the approval from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
* [ ] CI is green, or failing jobs are not related to this change (and you are 100% sure about this statement)
* [ ] A docker base image is modified?
    * [ ] the relevant `build-XXX-image` label is present
* [ ] A scenario is added (or removed)?
    * [ ] Get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

### CODE REVIEW

**Review by manuel-alvarez-alvarez — APPROVED** [review]
I'm missing some of the headers from [confluence](https://datadoghq.atlassian.net/wiki/spaces/APS/pages/2186870984/HTTP+header+collection) but looks good to me :smile:

**Comment by Anilm3** [comment]
> I'm missing some of the headers from [confluence](https://datadoghq.atlassian.net/wiki/spaces/APS/pages/2186870984/HTTP+header+collection) but looks good to me 😄

Indeed, I decided not to include all headers in this particular test as there are a few of them which aren't supported at all by most of the libraries (some XFF, edge WAF) and I wanted to ensure that the implementation of this feature isn't necessarily subject to exhaustive support of all headers, which should be tested separately.

File: tests/appsec/test_automated_login_events.py [hunk 1]
```diff
@@ -508,6 +528,52 @@ def test_login_sdk_failure_local(self):
...
+    @missing_feature(library="java")
+    @missing_feature(library="nodejs")
+    @missing_feature(library="python")
+    @missing_feature(library="php")
+    @missing_feature(library="ruby")
+    def test_login_success_headers(self):
+        # Validate that all relevant headers are included on user login success on extended mode
+
+        def validate_login_success_headers(span):
+            for header, _ in self.HEADERS.items():
```
Comment (cbeauchesne) [inline tests/appsec/test_automated_login_events.py:548]:
```suggestion
            for header in self.HEADERS:
```

File: tests/appsec/test_automated_login_events.py [hunk 2]
```diff
@@ -508,6 +528,52 @@ def test_login_sdk_failure_local(self):
...
+    @missing_feature(library="java")
+    @missing_feature(library="nodejs")
+    @missing_feature(library="python")
+    @missing_feature(library="php")
+    @missing_feature(library="ruby")
+    def test_login_failure_headers(self):
+        # Validate that all relevant headers are included on user login failure on extended mode
+
+        def validate_login_failure_headers(span):
+            for header, _ in self.HEADERS.items():
```
Comment (cbeauchesne) [inline tests/appsec/test_automated_login_events.py:571]:
```suggestion
            for header in self.HEADERS:
```

File: tests/appsec/test_automated_login_events.py (hunk 1 above)
Comment (cbeauchesne) [inline tests/appsec/test_automated_login_events.py:548]: I guess you are only validating the root span ? 

```suggestion
            if span.get("parent_id") not in (0, None):
                return
                
            for header, _ in self.HEADERS.items():
```

Yes, we miss some helper here...

File: tests/appsec/test_automated_login_events.py (hunk 2 above)
Comment (cbeauchesne) [reply tests/appsec/test_automated_login_events.py:571]: + same suggestion as above

File: tests/appsec/test_event_tracking.py [hunk 3]
```diff
@@ -37,6 +58,25 @@ def validate_user_login_success_tags(span):
...
+    @missing_feature(library="dotnet")
+    @missing_feature(library="java")
+    @missing_feature(library="nodejs")
+    @missing_feature(library="python")
+    @missing_feature(library="php")
+    @missing_feature(library="ruby")
+    def test_user_login_success_header_collection(self):
+        # Validate that all relevant headers are included on user login success
+
+        def validate_user_login_success_header_collection(span):
```
Comment (cbeauchesne) [inline tests/appsec/test_event_tracking.py:73]: Same two suggestions

File: tests/appsec/test_event_tracking.py [hunk 4]
```diff
@@ -72,6 +112,25 @@ def validate_user_login_failure_tags(span):
...
+    @missing_feature(library="dotnet")
+    @missing_feature(library="java")
+    @missing_feature(library="nodejs")
+    @missing_feature(library="python")
+    @missing_feature(library="php")
+    @missing_feature(library="ruby")
+    def test_user_login_failure_header_collection(self):
+        # Validate that all relevant headers are included on user login failure
+
+        def validate_user_login_failure_header_collection(span):
```
Comment (cbeauchesne) [inline tests/appsec/test_event_tracking.py:127]: Same two suggestions

**Review by cbeauchesne — CHANGES_REQUESTED** [review]
Just small suggestions. The rest is fine

File: tests/appsec/test_automated_login_events.py [hunk 5]
```diff
@@ -276,6 +275,27 @@ def password_key(self):
     BASIC_AUTH_USER_HEADER = "Basic dGVzdDoxMjM0"  # base64(test:1234)
     BASIC_AUTH_USER_UUID_HEADER = "Basic dGVzdHV1aWQ6MTIzNA=="  # base64(testuuid:1234)
 
+    HEADERS = {
```
[the remark this answers is not in the document]
Comment (Anilm3) [reply tests/appsec/test_automated_login_events.py:278]: If they were used more widely, this would certainly be the way to go, perhaps once all libraries report all headers we can just move them.

File: tests/appsec/test_automated_login_events.py (hunk 1 above)
Comment (Anilm3) [reply tests/appsec/test_automated_login_events.py]: Done

File: tests/appsec/test_automated_login_events.py (hunk 2 above)
Comment (Anilm3) [reply tests/appsec/test_automated_login_events.py]: Done

File: tests/appsec/test_event_tracking.py (hunk 4 above)
Comment (Anilm3) [reply tests/appsec/test_event_tracking.py:130]: Done

**Review by cbeauchesne — APPROVED** [review]
